File: backend/application/config/feature_flag_manager.py
"""Feature Flag Manager (Phase 14.3.4).

Thread-safe manager for feature flag registration, state toggling, evaluation with rollout percentages,
environment/profile restrictions, dependency flag resolution, and evaluation caching.
"""

from datetime import datetime, timezone
import hashlib
import logging
from threading import RLock
from typing import Dict, List, Optional, Set, Tuple

from backend.application.config.models import (
    FeatureEvaluation,
    FeatureFlag,
    FeatureHealth,
    FeatureStatistics,
)

logger = logging.getLogger(__name__)


class FeatureFlagManager:
    """Production thread-safe feature flag evaluation engine."""

    def __init__(self) -> None:
        """Initialize FeatureFlagManager."""
        self._lock = RLock()
        self._features: Dict[str, FeatureFlag] = {}
        self._evaluation_cache: Dict[str, FeatureEvaluation] = {}

        # Metrics
        self._evaluations_count: int = 0
        self._cache_hits: int = 0

    def register_feature(self, feature: FeatureFlag) -> bool:
        """Register or update a feature flag definition.

        Args:
            feature: Target FeatureFlag model.

        Returns:
            bool: True if registered.
        """
        with self._lock:
            self._features[feature.feature_name] = feature
            self._evaluation_cache.clear()
            logger.info("Registered feature flag '%s' (enabled=%s).", feature.feature_name, feature.enabled)
            return True

    def remove_feature(self, feature_name: str) -> bool:
        """Remove a registered feature flag.

        Args:
            feature_name: Target feature flag name.

        Returns:
            bool: True if removed.
        """
        with self._lock:
            if feature_name in self._features:
                del self._features[feature_name]
                self._evaluation_cache.clear()
                logger.info("Removed feature flag '%s'.", feature_name)
                return True
            return False

    def enable(self, feature_name: str) -> bool:
        """Enable a feature flag.

        Args:
            feature_name: Target feature flag name.

        Returns:
            bool: True if enabled.
        """
        with self._lock:
            if feature_name in self._features:
                flag = self._features[feature_name]
                if not flag.enabled:
                    self._features[feature_name] = FeatureFlag(
                        feature_name=flag.feature_name,
                        enabled=True,
                        description=flag.description,
                        rollout_percentage=flag.rollout_percentage,
                        allowed_profiles=flag.allowed_profiles,
                        allowed_environments=flag.allowed_environments,
                        dependencies=flag.dependencies,
                    )
                    self._evaluation_cache.clear()
                    logger.info("Enabled feature flag '%s'.", feature_name)
                return True
            return False

    def disable(self, feature_name: str) -> bool:
        """Disable a feature flag.

        Args:
            feature_name: Target feature flag name.

        Returns:
            bool: True if disabled.
        """
        with self._lock:
            if feature_name in self._features:
                flag = self._features[feature_name]
                if flag.enabled:
                    self._features[feature_name] = FeatureFlag(
                        feature_name=flag.feature_name,
                        enabled=False,
                        description=flag.description,
                        rollout_percentage=flag.rollout_percentage,
                        allowed_profiles=flag.allowed_profiles,
                        allowed_environments=flag.allowed_environments,
                        dependencies=flag.dependencies,
                    )
                    self._evaluation_cache.clear()
                    logger.info("Disabled feature flag '%s'.", feature_name)
                return True
            return False

    def toggle(self, feature_name: str) -> bool:
        """Toggle feature flag status.

        Args:
            feature_name: Target feature flag name.

        Returns:
            bool: Updated boolean status.
        """
        with self._lock:
            if feature_name in self._features:
                flag = self._features[feature_name]
                if flag.enabled:
                    self.disable(feature_name)
                    return False
                else:
                    self.enable(feature_name)
                    return True
            return False

    def _deterministic_rollout_score(self, feature_name: str, instance_id: str) -> float:
        """Compute deterministic rollout score (0.0 to 100.0) from MD5 hash."""
        seed = f"{feature_name}:{instance_id}".encode("utf-8")
        digest = hashlib.md5(seed).hexdigest()
        val = int(digest[:8], 16)
        return (val % 10000) / 100.0

    def evaluate(
        self,
        feature_name: str,
        active_profile_name: str = "development",
        active_env: str = "development",
        instance_id: str = "default",
    ) -> FeatureEvaluation:
        """Evaluate feature flag status with restrictions, dependencies, and rollout rules.

        Args:
            feature_name: Name of target feature.
            active_profile_name: Active profile name string.
            active_env: Active environment string.
            instance_id: Unique instance identifier for rollout hashing.

        Returns:
            FeatureEvaluation: Evaluation result model.
        """
        with self._lock:
            self._evaluations_count += 1
            cache_key = f"{feature_name}:{active_profile_name}:{active_env}:{instance_id}"

            if cache_key in self._evaluation_cache:
                self._cache_hits += 1
                return self._evaluation_cache[cache_key]

            # Missing feature flag check
            if feature_name not in self._features:
                eval_res = FeatureEvaluation(
                    feature_name=feature_name,
                    enabled=False,
                    reason=f"Feature flag '{feature_name}' is not registered.",
                    profile_name=active_profile_name,
                    environment_name=active_env,
                    evaluated_at=datetime.now(timezone.utc),
                )
                self._evaluation_cache[cache_key] = eval_res
                return eval_res

            flag = self._features[feature_name]

            # 1. Base enabled check
            if not flag.enabled:
                eval_res = FeatureEvaluation(
                    feature_name=feature_name,
                    enabled=False,
                    reason=f"Feature flag '{feature_name}' is explicitly disabled.",
                    profile_name=active_profile_name,
                    environment_name=active_env,
                    evaluated_at=datetime.now(timezone.utc),
                )
                self._evaluation_cache[cache_key] = eval_res
                return eval_res

            # 2. Dependencies check
            for dep in flag.dependencies:
                dep_eval = self.evaluate(dep, active_profile_name, active_env, instance_id)
                if not dep_eval.enabled:
                    eval_res = FeatureEvaluation(
                        feature_name=feature_name,
                        enabled=False,
                        reason=f"Required dependency feature flag '{dep}' is disabled.",
                        profile_name=active_profile_name,
                        environment_name=active_env,
                        evaluated_at=datetime.now(timezone.utc),
                    )
                    self._evaluation_cache[cache_key] = eval_res
                    return eval_res

            # 3. Allowed profiles check
            if flag.allowed_profiles:
                allowed_str_profiles = tuple(p.value.lower() for p in flag.allowed_profiles)
                if active_profile_name.lower() not in allowed_str_profiles:
                    eval_res = FeatureEvaluation(
                        feature_name=feature_name,
                        enabled=False,
                        reason=f"Profile '{active_profile_name}' is not in allowed profiles {allowed_str_profiles}.",
                        profile_name=active_profile_name,
                        environment_name=active_env,
                        evaluated_at=datetime.now(timezone.utc),
                    )
                    self._evaluation_cache[cache_key] = eval_res
                    return eval_res

            # 4. Allowed environments check
            if flag.allowed_environments:
                allowed_str_envs = tuple(e.lower() for e in flag.allowed_environments)
                if active_env.lower() not in allowed_str_envs:
                    eval_res = FeatureEvaluation(
                        feature_name=feature_name,
                        enabled=False,
                        reason=f"Environment '{active_env}' is not in allowed environments {allowed_str_envs}.",
                        profile_name=active_profile_name,
                        environment_name=active_env,
                        evaluated_at=datetime.now(timezone.utc),
                    )
                    self._evaluation_cache[cache_key] = eval_res
                    return eval_res

            # 5. Rollout percentage check
            if flag.rollout_percentage < 100.0:
                score = self._deterministic_rollout_score(feature_name, instance_id)
                if score > flag.rollout_percentage:
                    eval_res = FeatureEvaluation(
                        feature_name=feature_name,
                        enabled=False,
                        reason=f"Instance score ({score:.1f}%) exceeds rollout percentage ({flag.rollout_percentage}%).",
                        profile_name=active_profile_name,
                        environment_name=active_env,
                        evaluated_at=datetime.now(timezone.utc),
                    )
                    self._evaluation_cache[cache_key] = eval_res
                    return eval_res

            # All checks passed
            eval_res = FeatureEvaluation(
                feature_name=feature_name,
                enabled=True,
                reason="Feature flag active and all evaluation checks passed.",
                profile_name=active_profile_name,
                environment_name=active_env,
                evaluated_at=datetime.now(timezone.utc),
            )
            self._evaluation_cache[cache_key] = eval_res
            return eval_res

    def is_enabled(
        self,
        feature_name: str,
        active_profile_name: str = "development",
        active_env: str = "development",
        instance_id: str = "default",
    ) -> bool:
        """Check if feature flag is enabled."""
        eval_res = self.evaluate(feature_name, active_profile_name, active_env, instance_id)
        return eval_res.enabled

    def list_features(self) -> Tuple[FeatureFlag, ...]:
        """List all registered feature flags."""
        with self._lock:
            return tuple(self._features.values())

    def health(self) -> FeatureHealth:
        """Get health status of feature flag subsystem."""
        with self._lock:
            return FeatureHealth(
                is_healthy=True,
                issues=(),
                checked_at=datetime.now(timezone.utc),
            )

    def statistics(self) -> FeatureStatistics:
        """Get feature flag metrics."""
        with self._lock:
            enabled_count = sum(1 for f in self._features.values() if f.enabled)
            return FeatureStatistics(
                total_features=len(self._features),
                enabled_features=enabled_count,
                evaluations_count=self._evaluations_count,
                cache_hits=self._cache_hits,
            )

File: frontend/src/runtime/config/feature_flag_manager.ts
/**
 * Feature Flag Manager & Evaluation Engine (Phase 16.3.4).
 *
 * Implements feature flag registration, enable/disable/toggle controls, dependency evaluation,
 * profile & environment restrictions, deterministic rollout percentage calculation,
 * evaluation caching, and diagnostic reporting.
 */

import {
  createFeatureEvaluation,
  createFeatureFlag,
  createFeatureHealth,
  createFeatureStatistics,
  FeatureEvaluation,
  FeatureFlag,
  FeatureHealth,
  FeatureStatistics,
} from './models';
import { ConfigurationProviderException } from './exceptions';

export class FeatureFlagManager {
  private readonly _features = new Map<string, FeatureFlag>();
  private readonly _evaluationCache = new Map<string, FeatureEvaluation>();

  private _evaluations = 0;
  private _enabledEvaluations = 0;
  private _disabledEvaluations = 0;
  private _cachedEvaluations = 0;

  public registerFeature(feature: FeatureFlag): void {
    if (!feature) {
      throw new ConfigurationProviderException('Feature flag cannot be null or undefined.');
    }
    const name = feature.featureName.trim();
    if (!name) {
      throw new ConfigurationProviderException('Feature name cannot be empty.');
    }
    if (this._features.has(name)) {
      throw new ConfigurationProviderException(`Feature flag '${name}' is already registered.`);
    }

    this._features.set(name, feature);
    this._evaluationCache.clear();
  }

  public removeFeature(featureName: string): boolean {
    const name = featureName.trim();
    const removed = this._features.delete(name);
    if (removed) {
      this._evaluationCache.clear();
    }
    return removed;
  }

  public enable(featureName: string): void {
    const name = featureName.trim();
    const target = this._features.get(name);
    if (!target) {
      throw new ConfigurationProviderException(`Feature flag '${featureName}' is not registered.`);
    }
    this._features.set(name, createFeatureFlag({ ...target, enabled: true }));
    this._evaluationCache.clear();
  }

  public disable(featureName: string): void {
    const name = featureName.trim();
    const target = this._features.get(name);
    if (!target) {
      throw new ConfigurationProviderException(`Feature flag '${featureName}' is not registered.`);
    }
    this._features.set(name, createFeatureFlag({ ...target, enabled: false }));
    this._evaluationCache.clear();
  }

  public toggle(featureName: string): boolean {
    const name = featureName.trim();
    const target = this._features.get(name);
    if (!target) {
      throw new ConfigurationProviderException(`Feature flag '${featureName}' is not registered.`);
    }
    const nextState = !target.enabled;
    this._features.set(name, createFeatureFlag({ ...target, enabled: nextState }));
    this._evaluationCache.clear();
    return nextState;
  }

  public evaluate(
    featureName: string,
    context?: { profileName?: string; environmentName?: string; userId?: string },
  ): FeatureEvaluation {
    this._evaluations++;
    const name = featureName.trim();
    const cacheKey = `${name}:${context?.profileName ?? ''}:${context?.environmentName ?? ''}:${context?.userId ?? ''}`;

    if (this._evaluationCache.has(cacheKey)) {
      this._cachedEvaluations++;
      return this._evaluationCache.get(cacheKey)!;
    }

    const feature = this._features.get(name);
    if (!feature) {
      this._disabledEvaluations++;
      const res = createFeatureEvaluation({
        featureName: name,
        enabled: false,
        reason: 'Feature not registered.',
        profileName: context?.profileName,
        environmentName: context?.environmentName,
      });
      this._evaluationCache.set(cacheKey, res);
      return res;
    }

    if (!feature.enabled) {
      this._disabledEvaluations++;
      const res = createFeatureEvaluation({
        featureName: name,
        enabled: false,
        reason: 'Feature is explicitly disabled.',
        profileName: context?.profileName,
        environmentName: context?.environmentName,
      });
      this._evaluationCache.set(cacheKey, res);
      return res;
    }

    if (
      feature.allowedProfiles &&
      feature.allowedProfiles.length > 0 &&
      context?.profileName &&
      !feature.allowedProfiles.includes(context.profileName)
    ) {
      this._disabledEvaluations++;
      const res = createFeatureEvaluation({
        featureName: name,
        enabled: false,
        reason: `Profile '${context.profileName}' is not in allowed profiles.`,
        profileName: context?.profileName,
        environmentName: context?.environmentName,
      });
      this._evaluationCache.set(cacheKey, res);
      return res;
    }

    if (
      feature.allowedEnvironments &&
      feature.allowedEnvironments.length > 0 &&
      context?.environmentName &&
      !feature.allowedEnvironments.includes(context.environmentName)
    ) {
      this._disabledEvaluations++;
      const res = createFeatureEvaluation({
        featureName: name,
        enabled: false,
        reason: `Environment '${context.environmentName}' is not in allowed environments.`,
        profileName: context?.profileName,
        environmentName: context?.environmentName,
      });
      this._evaluationCache.set(cacheKey, res);
      return res;
    }

    if (feature.dependencies && feature.dependencies.length > 0) {
      for (const dep of feature.dependencies) {
        const depEval = this.evaluate(dep, context);
        if (!depEval.enabled) {
          this._disabledEvaluations++;
          const res = createFeatureEvaluation({
            featureName: name,
            enabled: false,
            reason: `Dependency requirement '${dep}' not met.`,
            profileName: context?.profileName,
            environmentName: context?.environmentName,
          });
          this._evaluationCache.set(cacheKey, res);
          return res;
        }
      }
    }

    if (feature.rolloutPercentage !== undefined && feature.rolloutPercentage < 100) {
      const userSeed = context?.userId ?? 'default_user';
      const hash = this.deterministicHash(userSeed);
      if (hash >= feature.rolloutPercentage) {
        this._disabledEvaluations++;
        const res = createFeatureEvaluation({
          featureName: name,
          enabled: false,
          reason: `User excluded by rollout percentage (${hash}% >= ${feature.rolloutPercentage}%).`,
          profileName: context?.profileName,
          environmentName: context?.environmentName,
        });
        this._evaluationCache.set(cacheKey, res);
        return res;
      }
    }

    this._enabledEvaluations++;
    const res = createFeatureEvaluation({
      featureName: name,
      enabled: true,
      reason: 'Feature enabled.',
      profileName: context?.profileName,
      environmentName: context?.environmentName,
    });
    this._evaluationCache.set(cacheKey, res);
    return res;
  }

  public listFeatures(): ReadonlyArray<FeatureFlag> {
    return Object.freeze(Array.from(this._features.values()));
  }

  public statistics(): FeatureStatistics {
    return createFeatureStatistics({
      evaluations: this._evaluations,
      enabledEvaluations: this._enabledEvaluations,
      disabledEvaluations: this._disabledEvaluations,
      cachedEvaluations: this._cachedEvaluations,
    });
  }

  public health(): FeatureHealth {
    const features = Array.from(this._features.values());
    const enabled = features.filter((f) => f.enabled).length;
    return createFeatureHealth({
      healthy: true,
      totalFeatures: features.length,
      enabledFeatures: enabled,
      disabledFeatures: features.length - enabled,
    });
  }

  public clear(): void {
    this._features.clear();
    this._evaluationCache.clear();
  }

  private deterministicHash(seed: string): number {
    let hash = 0;
    for (let i = 0; i < seed.length; i++) {
      hash = (hash << 5) - hash + seed.charCodeAt(i);
      hash |= 0;
    }
    return Math.abs(hash) % 100;
  }
}
